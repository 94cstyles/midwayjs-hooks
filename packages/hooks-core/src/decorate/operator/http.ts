import { DefineHelper, Operator, OperatorType } from '../type'

export const HttpTrigger = 'HTTP'

export enum HttpMetadata {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAM = 'Http_Param',
  HEADER = 'Http_Header',
  RESPONSE = 'Http_Response',
}

export enum ResponseMetadata {
  CODE = 'Http_Response_Code',
  HEADER = 'Http_Response_Header',
  CONTENT_TYPE = 'Http_Response_ContentType',
  REDIRECT = 'Http_Response_Redirect',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      metadata({ setMetadata }) {
        setMetadata(OperatorType.Trigger, {
          type: HttpTrigger,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Delete = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpMetadata.QUERY,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpMetadata.PARAM,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpMetadata.HEADER,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.HEADER, true)
    },
  }
}

export type ResponseMetaData = {
  type: ResponseMetadata
  code?: number
  header?: {
    key: string
    value: string
  }
  url?: string
  contentType?: string
}

export function HttpCode(code: number): Operator<void> {
  return {
    name: 'HttpCode',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.CODE, { code })
    },
  }
}

export function SetHeader(key: string, value: string): Operator<void> {
  return {
    name: 'SetHeader',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.HEADER, {
        header: {
          key,
          value,
        },
      })
    },
  }
}

export function Redirect(url: string, code?: number): Operator<void> {
  return {
    name: 'Redirect',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.REDIRECT, {
        url,
        code,
      })
    },
  }
}

export function ContentType(contentType: string): Operator<void> {
  return {
    name: 'ContentType',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.CONTENT_TYPE, {
        contentType,
      })
    },
  }
}

function setResponseMetaData(
  helper: DefineHelper,
  type: ResponseMetadata,
  value: Partial<ResponseMetaData>
) {
  const responseMetaData =
    helper.getMetadata<ResponseMetaData[]>(HttpMetadata.RESPONSE) || []

  helper.setMetadata(HttpMetadata.RESPONSE, [
    ...responseMetaData,
    {
      type,
      ...value,
    },
  ])
}
