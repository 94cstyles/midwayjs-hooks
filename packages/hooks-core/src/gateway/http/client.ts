import { EXPORT_DEFAULT_FUNCTION_ALIAS } from '../../const'
import { createApiClientMatcher } from '../../index'
import { HTTPRouter } from './router'

function getUrl(baseUrl: string, name: string) {
  if (name === EXPORT_DEFAULT_FUNCTION_ALIAS) {
    return baseUrl
  }

  if (baseUrl.endsWith('/')) {
    return `${baseUrl}${name}`
  }

  return `${baseUrl}/${name}`
}

export function createHTTPClientMatcher(router: HTTPRouter) {
  return createApiClientMatcher().match((meta) => {
    const { file, route, functionName } = meta

    if (!route.basePath) {
      return
    }

    const baseUrl = router.getBaseUrl(file)
    const url = getUrl(baseUrl, functionName)
    const client =
      router.projectConfig?.request?.client || '@midwayjs/hooks/request'

    return {
      client,
      metadata: {
        url,
        meta: {},
      },
    }
  })
}
