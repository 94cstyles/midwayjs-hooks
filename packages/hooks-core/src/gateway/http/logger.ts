import { relative } from 'upath'

export function duplicateLogger(
  root: string,
  existPath: string,
  currentPath: string,
  api: string
) {
  console.log(
    '[%s] Duplicate routes detected. %s and %s both resolve to %s.',
    'warn',
    relative(root, existPath),
    relative(root, currentPath),
    api
  )
}
