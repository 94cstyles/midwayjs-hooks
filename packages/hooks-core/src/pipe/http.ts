import { Operator, OperatorProperty } from './type'

const Protocol = 'HTTP'

enum HttpProperty {
  METHOD = 'METHOD',
  QUERY = 'QUERY',
  PARAM = 'PARAM',
  HEADER = 'HEADER',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      defineMeta({ setProperty: defineProperty }) {
        defineProperty(OperatorProperty.Trigger, {
          type: Protocol,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Delete = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpProperty.QUERY,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpProperty.PARAM,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpProperty.HEADER,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.HEADER, true)
    },
  }
}
