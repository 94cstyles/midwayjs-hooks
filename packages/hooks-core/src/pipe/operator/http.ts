import { Operator, OperatorType } from '../type'

export const HttpTrigger = 'HTTP'

export enum HttpProperty {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAM = 'Http_Param',
  HEADER = 'Http_Header',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      defineMeta({ setProperty }) {
        setProperty(OperatorType.Trigger, {
          type: HttpTrigger,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Del = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpProperty.QUERY,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpProperty.PARAM,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpProperty.HEADER,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.HEADER, true)
    },
  }
}
