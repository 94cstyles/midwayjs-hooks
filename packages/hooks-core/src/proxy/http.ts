import { proxy } from './proxy'
import { request } from '../request'
import { getHTTPMethod } from '../router'

export function createRequestProxy(baseUrl: string) {
  return proxy(async (func, thisArg, args) => {
    const name = func._param.meta.name
    return request({
      url: baseUrl + name,
      method: getHTTPMethod(args.length),
      data: { args },
      meta: {},
    })
  })
}
