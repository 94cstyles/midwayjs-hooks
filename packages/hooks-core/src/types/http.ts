/**
 * Now the function accepts all http methods
 * @deprecated will be removed in next major version.
 */
export type ApiHttpMethod = 'GET' | 'POST'

export type ApiBaseParam = {
  /**
   * @description generated by the file router, base on file path
   */
  functionId?: string
}

export interface ApiParam extends ApiBaseParam {
  url: string
  /**
   * Now the function accepts all http methods
   * @deprecated will be removed in next major version.
   */
  method: ApiHttpMethod
  data: {
    args: any[]
    [key: string]: any
  }
  meta: {
    functionName?: string
    superjson?: boolean
    [key: string]: any
  }
}
