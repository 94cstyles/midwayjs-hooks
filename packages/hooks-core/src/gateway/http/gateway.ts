import { All, Controller } from '@midwayjs/decorator'

import { useApiClientMatcher } from '../../request/builder'
import { als, createFunctionContainer } from '../../runtime'
import { Route } from '../../types/config'
import {
  ComponentOptions,
  CreateApiOptions,
  GatewayAdapterOptions,
  HooksGatewayAdapter,
  OnReadyArgs,
} from '../../types/gateway'
import { useHooksMiddleware } from '../../util'
import { createHTTPClientMatcher } from './client'
import { HTTPRouter } from './router'

export class HTTPGateway implements HooksGatewayAdapter {
  options: GatewayAdapterOptions

  private readonly router: HTTPRouter

  is(route: Route) {
    return !!route?.basePath
  }

  constructor(options: ComponentOptions) {
    this.options = options
    this.router = new HTTPRouter(options)

    useApiClientMatcher(createHTTPClientMatcher(this.router))
  }

  createApi(options: CreateApiOptions) {
    const { file, functionName, isExportDefault, fn } = options
    const httpPath = this.router.getHTTPPath(
      file,
      functionName,
      isExportDefault
    )

    fn._param.url = httpPath
    this.createHTTPApi(httpPath, options)
  }

  createHTTPApi(httpPath: string, options: CreateApiOptions) {
    const { functionId, fn } = options
    // setup middleware
    const middleware = [...fn.middleware]

    createFunctionContainer({
      runWithAsyncLocalStorage: false,
      fn,
      functionId,
      parseArgs(ctx) {
        return ctx.request?.body?.args || []
      },
      classDecorators: [Controller(httpPath)],
      handlerDecorators: [All('/', { middleware })],
    })
  }

  async onReady(args: OnReadyArgs) {
    const { app, runtimeConfig } = args
    const mws = [this.useAsyncLocalStorage]
    runtimeConfig.middleware?.forEach?.((mw) =>
      mws.push(useHooksMiddleware(mw))
    )

    for (const mw of mws) {
      ;(app as any).use(mw)
    }
  }

  useAsyncLocalStorage = async (ctx: any, next: any) => {
    await als.run({ ctx }, async () => await next())
  }
}
