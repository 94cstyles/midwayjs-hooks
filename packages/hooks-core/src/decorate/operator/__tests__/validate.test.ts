import { z } from 'zod'
import { Decorate } from '../..'
import { Post } from '../http'
import { Validate, Validator, setValidator } from '../validate'

describe('validator', () => {
  it('use default validator', async () => {
    const schema = z.number()

    const fn = Decorate(
      Post(),
      Validate(schema),
      async (count: z.infer<typeof schema>) => {
        return count
      }
    )

    expect(await fn(1)).toEqual(1)
  })

  it('should throw error', async () => {
    const schema = z.object({
      name: z.string(),
      age: z.number(),
    })

    const fn = Decorate(
      Post(),
      Validate(schema),
      async (user: z.infer<typeof schema>) => {
        return user
      }
    )

    expect(fn({ name: 'hooks', age: 1 })).toBeTruthy()
    expect(
      fn({ name: 1, age: 1 } as any)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(
      fn({ name: '', age: '' } as any)
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should support custom validator', async () => {
    const validator: Validator = (schema, input) => {
      if (schema === String) {
        if (typeof input !== 'string') {
          throw new Error('input must be a string')
        }
      } else {
        throw new Error('unsupported schema')
      }
    }

    setValidator(validator)

    const fn = Decorate(Post(), Validate(String), async (name: string) => {
      return { name }
    })

    expect(await fn('hooks')).toEqual({ name: 'hooks' })
    expect(fn(1 as any)).rejects.toThrowErrorMatchingSnapshot()

    const fn2 = Decorate(Post(), Validate(Number), async (age: number) => {
      return { age }
    })
    expect(fn2(1)).rejects.toThrowErrorMatchingSnapshot()
  })
})
