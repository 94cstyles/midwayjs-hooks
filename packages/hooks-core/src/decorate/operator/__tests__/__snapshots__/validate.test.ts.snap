// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator should support custom validator 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"number\\",
    \\"path\\": [
      \\"name\\"
    ],
    \\"message\\": \\"Expected string, received number\\"
  }
]"
`;

exports[`validator should support custom validator 2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"string\\",
    \\"path\\": [
      \\"age\\"
    ],
    \\"message\\": \\"Expected number, received string\\"
  }
]"
`;

exports[`validator should support custom validator 3`] = `"input must be a string"`;

exports[`validator should support custom validator 4`] = `"unsupported schema"`;
